### PR TITLE
Added GaussianBlur transform

### PR DESCRIPTION
**Pull request by tejank10**
This PR addresses issue #[2635](https://github.com/pytorch/vision/issues/2635). It adds GaussianBlur image transformation. It takes a range of sigma as input. It picks a sigma for Gaussian kernel at random and applies the kernel to the image. 

- [x] Implement the transform
- [x] Add tests
- [x] Add docs

Let me know your thoughts @vfdev-5, @yaox12  on where could it be possibly improved.

### CODE REVIEW

**Comment by oke-aditya** [comment]
Small linting errors.

```
./torchvision/transforms/functional_tensor.py:473:16: E225 missing whitespace around operator
./torchvision/transforms/functional_tensor.py:482:23: E221 multiple spaces before operator
./torchvision/transforms/functional_tensor.py:483:22: E226 missing whitespace around arithmetic operator
./torchvision/transforms/functional_tensor.py:1001:31: E226 missing whitespace around arithmetic operator
./torchvision/transforms/functional_tensor.py:1002:19: E221 multiple spaces before operator
```

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -1545,3 +1544,62 @@ def forward(self, img):
...
+    Args:
+        rad_min (float): Minimum radius that can be chosen for blurring kernel.
+        rad_max (float): Maximum radius that can be chosen for blurring kernel.
+
+    Returns:
+        PIL Image or Tensor: Gaussian blurred version of the input image.
+
+    """
+
+    def __init__(self, rad_min=0.1, rad_max=2.0):
```
Comment (yaox12) [inline torchvision/transforms/transforms.py:1564]: I recommend the input to be consistent with those of `ColorJitter`, i.e., `float or tuple of float (min, max)`.
https://github.com/pytorch/vision/blob/a4736ea6e8ff25a1265ca5adf9e6e244d78500e8/torchvision/transforms/transforms.py#L1013-L1019

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -1545,3 +1544,62 @@ def forward(self, img):
...
+        """Choose radius for ``gaussian_blur`` for random gaussian blurring.
+
+        Args:
+            rad_min (float): Minimum radius that can be chosen for blurring kernel.
+            rad_max (float): Maximum radius that can be chosen for blurring kernel.
+
+        Returns:
+            float: radius be passed to ``gaussian_blur`` for gaussian blurring.
+        """
+        radius = torch.rand(1).item() * (rad_max - rad_min) + rad_min
```
Comment (yaox12) [inline torchvision/transforms/transforms.py:1590]: I think it may be better to use
```python
radius = random.uniform(rad_min, rad_max)
```

File: torchvision/transforms/transforms.py (hunk 1 above)
Comment (tejank10) [reply torchvision/transforms/transforms.py:1564]: If the `radius` is passed as `float`, does that imply we are going to choose a kernel with that fixed radius every time? Or would it mean we will choose a radius randomly between, (1-radius, 1+radius), like in `ColorJitter`?
The former one seems more intuitive to me because unlike brightness (and others in `ColorJitter`) there is no "central value"(=1 in `ColorJitter`) of radius here which indicates that it is the default one. What do you think?

Comment (yaox12) [reply torchvision/transforms/transforms.py:1564]: I think a fixed radius would be fine if it is passed as float.

**Comment by vfdev-5** [comment]
@tejank10 thanks for the PR and sorry for delay.

Finally I think the API we would like is with `kernel_size` and `sigma`:
```python
# functional.py
def gaussian_blur(img: Tensor, kernel_size: int, sigma: float) -> Tensor:
```
The main problem is with PIL which can apply kernels larger than 5x5. Maybe, a solution can be to reuse torch implementation for that : `PIL -> np.array -> torch.Tensor -> F_t.gaussian_blur -> np.array -> PIL`

About your implementation, I'll comment the things where I have questions.

File: torchvision/transforms/functional_tensor.py [hunk 3]
```diff
@@ -917,3 +940,70 @@ def perspective(
...
+
+    kernel = _get_kernel(radius, passes)
+
+    padding = _compute_padding(kernel.shape[::-1])
+
+    kernel = kernel[None, None, :, :].repeat(img.size(-3), 1, 1, 1)
+
+    padded_img = pad(img, padding, padding_mode='edge')
+    blurred_img = conv2d(padded_img, kernel, groups=img.size(-3))
+    for _ in range(passes - 1):
```
Comment (vfdev-5) [inline torchvision/transforms/functional_tensor.py:1001]: Why do we use `passes` actually ?

Comment (tejank10) [reply torchvision/transforms/functional_tensor.py:1001]: I was trying to match PIL's gaussian blur behaviour. In PIL they perform Box filter `passes` (= 3) times to approximate the required Gaussian blur.

**Comment by vfdev-5** [comment]
@tejank10 OK, let's work on this PR the following way:

- implement `F_t.gaussian_blur(img: Tensor, kernel_size: int, sigma: float)` which should be close to opencv's implementation : kernel creation, padding, etc (maybe, [this](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/smooth.dispatch.cpp) is how it is done in opencv).
- implement tests

I propose you to work on `F_t.gaussian_blur` if you do not mind and I'll work on tests and commit my modifications directly to this PR. How does it sound ?

**Comment by tejank10** [comment]
Sounds good, will update the API.

**Comment by vfdev-5** [comment]
@yaox12 how about this API ?

File: torchvision/transforms/functional_tensor.py [hunk 4]
```diff
@@ -1046,71 +1023,60 @@ def _get_gaussian_kernel1d(kernel_size: int, sigma: float):
...
-    ksize_x, ksize_y = kernel_size
-    sigma_x, sigma_y = sigma
-
-    kernel1d_x = _get_gaussian_kernel1d(ksize_x, sigma_x)
-    kernel1d_y = _get_gaussian_kernel1d(ksize_y, sigma_y)
+def _get_gaussian_kernel2d(
+        kernel_size: List[int], sigma: List[float], dtype: torch.dtype, device: torch.device
+) -> Tensor:
+    kernel1d_x = _get_gaussian_kernel1d(kernel_size[1], sigma[0]).to(device, dtype=dtype)
+    kernel1d_y = _get_gaussian_kernel1d(kernel_size[0], sigma[1]).to(device, dtype=dtype)
```
Comment (tejank10) [inline torchvision/transforms/functional_tensor.py:1030]: Should it be `kernel_size[0]` for `kernel1d_x` and `kernel_size[1]` for `kernel1d_y`?

Comment (vfdev-5) [reply torchvision/transforms/functional_tensor.py:1030]: Yes, you are right, I was thinking to switch the convention but then decided to keep it as `(kx, ky)`. Can you fix that please ?

**Comment by vfdev-5** [comment]
@tejank10 we have to improve coverage by executing all input checkings and asserting that they raise correct exception. Can you work on that too ?

File: torchvision/transforms/functional_tensor.py [hunk 5]
```diff
@@ -1011,3 +1011,72 @@ def perspective(
...
+    out_dtype = img.dtype
+    need_cast = False
+    if out_dtype != kernel.dtype:
+        need_cast = True
+        img = img.to(kernel)
+
+    # padding = (left, right, top, bottom)
+    padding = [kernel_size[0] // 2, kernel_size[0] // 2, kernel_size[1] // 2, kernel_size[1] // 2]
+    img = torch_pad(img, padding, mode="reflect")
+    img = conv2d(img, kernel, groups=img.shape[-3])
```
Comment (vfdev-5) [inline torchvision/transforms/functional_tensor.py:1073]: Note: maybe we can try to use padding mode of conv2d, see https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#torch.nn.Conv2d

**Comment by vfdev-5** [comment]
@tejank10 do you plan to work on this PR this week ?  Such that we do not do the same things without syncing. Otherwise, I'll rework everything tomorrow.

**Comment by tejank10** [comment]
Yeah, I'll be working on this this week.

**Comment by vfdev-5** [comment]
@tejank10 could you please push all your work today if any. I'll finalize today / tomorrow this PR.

File: test/test_transforms.py [hunk 6]
```diff
@@ -1667,6 +1667,91 @@ def test_random_grayscale(self):
         # Checking if RandomGrayscale can be printed as string
         trans3.__repr__()
 
+    def test_random_erasing(self):
```
Comment (vfdev-5) [inline test/test_transforms.py:1670]: @tejank10 I think there was a problem while resolving the conflict with master

Comment (tejank10) [reply test/test_transforms.py:1670]: Okay, I'll figure it out.

**Comment by vfdev-5** [comment]
@tejank10 I'll fix broken CI and merge conflict today. Let me know if you want to push some other updates. Thanks

File: torchvision/transforms/functional_tensor.py (hunk 5 above)
Comment (vfdev-5) [reply torchvision/transforms/functional_tensor.py:1166]: Padding mode is not present as argument in F.conv2d and nn.Conv2d does internally F.pad + F.conv2d same as we do here.

Comment (tejank10) [reply torchvision/transforms/functional_tensor.py:1166]: Is it 'reflect' padding that they do or 0 padding? I suppose it is the latter one.

**Review by fmassa — APPROVED** [review]
Thanks a lot @tejank10 and @vfdev-5 !
